crates/ruff_linter/src/rules/flake8_pyi/rules/custom_type_var_return_type.rs

```diff
@@ -230,3 +236,183 @@ fn is_likely_private_typevar(type_var_name: &str, type_params: Option<&TypeParam
+        Fix::safe_edits(first, rest)
+    } else {
+        Fix::unsafe_edits(first, rest)
+    }
+}
+
+fn import_self(checker: &Checker, return_range: TextRange) -> Option<Edit> {
+    // From PYI034's fix
+    let target_version = checker.settings.target_version.as_tuple();
+    let source_module = if target_version >= (3, 11) {
+        "typing"
+    } else {
+        "typing_extensions"
+    };
+
+    let (importer, semantic) = (checker.importer(), checker.semantic());
+    let request = ImportRequest::import_from(source_module, "Self");
+
+    let position = return_range.start();
+    let (edit, ..) = importer
+        .get_or_import_symbol(&request, position, semantic)
+        .ok()?;
+
+    Some(edit)
+}
+
+fn remove_first_parameter_annotation(parameters: &Parameters) -> Edit {
+    let mut non_variadic_positional = parameters.posonlyargs.iter().chain(&parameters.args);
+    let first = &non_variadic_positional.next().unwrap().parameter;
+
+    let name_end = first.name.range.end();
+    let annotation_end = first.range.end();
+
+    Edit::deletion(name_end, annotation_end)
+}
+
+fn replace_return_annotation_with_self(returns: &Expr) -> Edit {
+    Edit::range_replacement("Self".to_string(), returns.range())
+}
+
+fn replace_typevar_usages_with_self(
+    parameters: &Parameters,
+    typevar_name: &str,
+) -> Result<Vec<Edit>, Vec<Edit>> {
+    let mut edits = vec![];
+    let mut could_not_handle_all_usages = false;
+
+    for parameter in parameters.iter().skip(1) {
+        let Some(annotation) = parameter.annotation() else {
+            continue;
+        };
+        let Expr::Name(name) = annotation else {
+            continue;
+        };
+
+        if name.id.as_str() == typevar_name {
+            let edit = Edit::range_replacement("Self".to_string(), annotation.range());
+            edits.push(edit);
+        } else {
+            // Bail for complex cases
+            could_not_handle_all_usages = true;
+        }
+    }
+
+    if could_not_handle_all_usages {
+        Err(edits)
+    } else {
+        Ok(edits)
+    }
+}
+
+fn remove_typevar_declaration(type_params: Option<&TypeParams>, name: &str) -> Option<Edit> {
+    let is_declaration_in_question = |type_param: &&TypeParam| -> bool {
+        if let TypeParam::TypeVar(typevar) = type_param {
+            return typevar.name.as_str() == name;
+        };
+
+        false
+    };
+
+    let parameter_list = type_params?;
+    let parameters = &parameter_list.type_params;
+    let first = parameters.first()?;
+
+    if parameter_list.len() == 1 && is_declaration_in_question(&first) {
+        return Some(Edit::range_deletion(parameter_list.range));
+    }
+
+    let (index, ..) = parameters
+        .iter()
+        .find_position(is_declaration_in_question)?;
+
+    let typevar_range = parameters[index].as_type_var().unwrap().range();
+    let last_index = parameters.len() - 1;
+
+    let range = if (0..last_index).contains(&index) {
```

Comment:
```suggestion
    let range = if index < last_index {
```

or should this include the last index? 

```
let range = if index <= last_index {